torchdata/datapipes/iter/util/header.py

```diff
@@ -20,14 +21,26 @@ class HeaderIterDataPipe(IterDataPipe[T_co]):
     def __init__(self, source_datapipe: IterDataPipe[T_co], limit: int = 10) -> None:
         self.source_datapipe: IterDataPipe[T_co] = source_datapipe
         self.limit: int = limit
+        self.length = -1
 
     def __iter__(self) -> Iterator[T_co]:
         for i, value in enumerate(self.source_datapipe):
             if i < self.limit:
                 yield value
             else:
                 break
+        self.length = min(i + 1, self.limit)  # We know length with certainty when we reach here
 
-    # TODO(134): Fix the case that the length of source_datapipe is shorter than limit
     def __len__(self) -> int:
-        return self.limit
+        if self.length != -1:
+            return self.length
+        try:
+            source_len = len(self.source_datapipe)
+            self.length = min(source_len, self.limit)
+            return self.length
+        except TypeError:
+            warn(
```

Comment:
Is raising a warning and returning the best guess (i.e. `self.limit`) the right decision here? It can be triggered by things like `list(header_dp)` if `header_dp` hasn't been fully traversed once yet, which might confuse the users. The alternative is to raise an exception and not return anything.

Reply:
This looks good to me.